### PR TITLE
[Merged by Bors] - Add `World::clear_resources` & `World::clear_all`

### LINKED ISSUE #3158: Add `World::clear_resources`
## What problem does this solve or what need does it fill?

`World::clear_entities` allows users to clear all entities. See
https://github.com/bevyengine/bevy/blob/f4776f2ec413bb80ff4ee51e04bb57f0a44d9983/crates/bevy_ecs/src/world/mod.rs#L1139

Clearing all resources is a natural parallel, and likely to be useful for various testing purposes.

## What solution would you like?

Add `World::clear_resources`.

Optionally, add a `clear_all` method that simply wraps `clear_entities` and `clear_resources`.

## What alternative(s) have you considered?

Awful manual iteration or type registration.

Creating a new world, moving all entities into it, then swapping them back?

## Additional context

Mechanics of `clear_entities` clarified by #3157.

### PR DESCRIPTION
**Pull request by 2ne1ugly**
# Objective

- Fixes #3158

## Solution

- clear columns

## Change Log

- Added `SparseSet::clear`
- Added `Resources::clear`
- Added `World::clear_resources`
- Added `World::clear_all`

My implementation of `clear_resources` do not remove the components itself but it clears the columns that keeps the resource data. I'm not sure if the issue meant to clear all resources, even the components and component ids (which I'm not sure if it's possible)

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
Looks good to me :) I think it's fine that we don't remove the columns themselves; there should be very little appreciable difference to the end user.

**Comment by alice-i-cecile** [comment]
Coming back to this now that I have merge powers: docs would be nice here, but they're also missing for `clear_entities`, so I won't block on it.

File: crates/bevy_ecs/src/world/mod.rs
Comment (MrGVSV) [inline crates/bevy_ecs/src/world/mod.rs:1155]: I imagine this clears internal Bevy resources as well? Should we document that? I think it could be an easy footgun for those who don't understand it.

File: crates/bevy_ecs/src/world/mod.rs
Comment (MrGVSV) [inline crates/bevy_ecs/src/world/mod.rs:1151]: Can we add some documentation to these methods?

File: crates/bevy_ecs/src/world/mod.rs
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/world/mod.rs:1155]: Yeah I'm sold we should document all three of these methods before merging.

They're pretty powerful footguns, and mostly useful for multi-world stuff or testing.

**Comment by alice-i-cecile** [comment]
@2ne1ugly if you're around, adding docs would be great. Otherwise I'll redo this PR in a week or two with docs (and credit to your work).

**Comment by 2ne1ugly** [comment]
> @2ne1ugly if you're around, adding docs would be great. Otherwise I'll redo this PR in a week or two with docs (and credit to your work).

I'll try adding them by friday

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_ecs/src/world/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/mod.rs:1278]:
```suggestion
    /// invalidating all [`Entity`] and resource fetches such as [`Res`](crate::system::Res), [`ResMut`](crate::system::ResMut)
```

File: crates/bevy_ecs/src/world/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/mod.rs:1284]:
```suggestion
    /// Despawns all entities in this [`World`].
```

File: crates/bevy_ecs/src/world/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/mod.rs:1285]:
```suggestion
```

I don't think this line adds anything

File: crates/bevy_ecs/src/world/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/mod.rs:1293]:
```suggestion
    /// Clears all resources in this [`World`].
```

File: crates/bevy_ecs/src/world/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/mod.rs:1294]:
```suggestion
    ///
    /// Any resource fetch to this [World] will fail unless they are re-initialized.
```

Line break between summary and details by convention.

**Comment by alice-i-cecile** [comment]
Apologies; the approval was in error. There are some doc comment fixes that should be addressed before merging.

**Review by alice-i-cecile — APPROVED** [review]

**Review by rodolphito — APPROVED** [review]

File: crates/bevy_ecs/src/world/mod.rs
Comment (james7132) [inline crates/bevy_ecs/src/world/mod.rs:1294]:
```suggestion
    /// **Note:** Any resource fetch to this [World] will fail unless they are re-initialized,
    /// including engine-internal resources that are only initialized on app/world construction.
    ///
    /// This can easily cause systems expecting certain resources to immediately start panicking.
    /// Use with caution.
```

I don't think this warning is strongly worded enough.

**Review by james7132 — CHANGES_REQUESTED** [review]

**Comment by alice-i-cecile** [comment]
I think this needs to be rebased to ensure that it compiles with the latest `main` :)

File: crates/bevy_ecs/src/world/mod.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/world/mod.rs:1627]: Typo here, clear entries is listed twice.

**Review by alice-i-cecile — APPROVED** [review]
bors r+

**Comment by james7132** [comment]
@2ne1ugly can you provide a changelog section to detail the changes in this PR?

**Comment by 2ne1ugly** [comment]
> @2ne1ugly can you provide a changelog section to detail the changes in this PR? If you want an example of how to write one, take a look at #7161.

Done

**Pull request closed without merging**
